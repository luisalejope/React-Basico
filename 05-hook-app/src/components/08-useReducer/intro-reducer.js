/* REGLAS DE LOS REDUCER
1. NO DEBE DE TENER EFECTOS SECUNDARIOS
2. NO DEBE DE REALIZAR TAREAS ASINCRONAS
3. DEBE DE RETORNAR SIEMPRE UN ESTADO NUEVO
4. NO DEBE DE LLAMAR LOCALSTORAGE O SESSIONSTORAGE
5. NO DEBE DE REQUERIR MAS QUE UNA ACCION QUE PUEDE REQUERIR UN ARGUMENTO
*/

// estado inicial
const initialState = [
  {
    id: 1,
    todo: "Comprar leche",
    done: false,
  },
];

/* 
recibe el estado y la accion
donde la aciion es el que va a modificar el estado
*/
const todoReducer = (state = initialState, action) => {

    if (action?.type === 'agregar') {
        return [...state, action.payload]
    }

  return state;
};

// reasigna los todos
let todos = todoReducer();


// este es el cambio que se necesita en el state
const newTodo = [
  {
    id: 2,
    todo: "Comprar pan",
    done: false,
  },
];

// esta es la accion para el reducer
const agregarTodoAction = {
    type: 'agregar',
    payload: newTodo,

}


todos = todoReducer(todos, agregarTodoAction)

console.log(todos);
