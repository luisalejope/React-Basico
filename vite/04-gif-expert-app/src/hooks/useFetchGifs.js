import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";


export const useFetchGifs = (category) => {
    
    const [gifs, setGifs] = useState([])
    const [isLoading, setIsLoading] = useState(false)

    const requestGifs = async() => {
        try {
            setIsLoading(true)
            const res = await getGifs(category)
            setGifs(res)
            console.log(res)
        } catch (error) {
            console.error(error)
        } finally {
            setIsLoading(false)
        }
    }

    useEffect(() => {
        requestGifs()
    }, [category])
    

    return {
        gifs,
        isLoading
    }
}
