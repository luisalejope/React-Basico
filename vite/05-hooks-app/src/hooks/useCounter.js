import { useState } from "react"


export const useCounter = (initialValue = 10) => {

    const [counter, setCounter] = useState(initialValue)

    const increase = (value = 1) => {
        setCounter(current => current + value)
    }

    const decrease = (value = 1) => {
        if (counter === 0) return;
        setCounter(current => current - value)
    }

    const reset = () => {
        setCounter(initialValue)
    }

    return {
        counter,
        increase,
        reset,
        decrease,
    }
}
