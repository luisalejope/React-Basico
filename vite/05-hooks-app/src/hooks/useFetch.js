import { useEffect, useState } from "react"

export const useFetch = (url) => {
    const [state, setState] = useState({
        data: null,
        isLoading: true,
        hasError: false,
        error: null
    })

    useEffect(() => {
        getFetch()
    }, [url])

    const setLoadingState = () => {
        setState({
            data: null,
            isLoading: true,
            hasError: false,
            error: null
        })
    }


    const getFetch = async()=> {

        setLoadingState();
        
        const response = await fetch(url);

        if (!response.ok) {
            setState({
                data: null,
                isLoading: false,
                hasError: true,
                error: {
                    code: response.status,
                    message: response.statusText
                }
            })
        }
        
        const data = await response.json()

        setState({
            data: data,
            isLoading: false,
            hasError: false,
            error: null
        })
        console.log(data)
    }
    


    return{
        data: state.data,
        isLoading: state.isLoading,
        hasError: state.hasError,
    }
}
