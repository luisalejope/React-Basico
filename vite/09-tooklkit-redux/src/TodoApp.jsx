import { useState } from "react"
import { useGetTodoByIdQuery, useGetTodosQuery } from "./store/apis/todosApi"


export const TodoApp = () => {
    const [todoId, setTodoId] = useState(1)
    // const {data: todos = [], isLoading} = useGetTodosQuery()
    const {data: todo, isLoading} = useGetTodoByIdQuery(todoId)

    const nextTodo = () =>{
        setTodoId(todoId + 1)
    }

    const prevTodo = () =>{
        if(todoId === 1) return;

        setTodoId(todoId - 1)
    }

  return (
    <>
        <h1>Todos - RTK QUERY</h1>
        <hr />
        <h4>Is Loading {isLoading ? 'true': 'false'} </h4>

        <pre>{JSON.stringify(todo)}</pre>

        {/* <ul>
            {
                todos.map(todo => (
                    <li key={todo.id}>
                        <strong>{todo.completed ? 'done' : 'completed'}</strong>
                        {todo.title}
                    </li>
                ))
            }
        </ul> */}

        <button onClick={prevTodo}>
            Prev Todo
        </button>
        <button onClick={nextTodo}>
            Next Todo
        </button>
    </>
  )
}
