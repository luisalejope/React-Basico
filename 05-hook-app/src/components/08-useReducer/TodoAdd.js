import React from "react";
import PropTypes from "prop-types";
import { useForm } from "../../hooks/useForm";

const TodoAdd = ({addTodo}) => {
    const [{ description }, handleInputChange, reset] = useForm({
        description: "",
      });
      const handleSubmit = (e) => {
        e.preventDefault();
    
        if (description.trim().length <= 1) {
          return;
        }
    
        // nuevo cambio en el estado
        const newTodo = {
          id: new Date().getTime(),
          desc: description,
          done: false,
        };
        addTodo(newTodo)
        reset();
      };
  return (
    <>
      <h4>Agregar</h4>
      <br />
      <form className="agregarForm" onSubmit={handleSubmit}>
        <input
          type="text"
          name="description"
          className="form-control"
          placeholder="Nuevo Todo"
          autoComplete="off"
          value={description}
          onChange={handleInputChange}
        />
        <button type="submit" className="btn btn-outline-primary mt-1">
          Agregar
        </button>
      </form>
    </>
  );
};

TodoAdd.propTypes = {};

export default TodoAdd;
