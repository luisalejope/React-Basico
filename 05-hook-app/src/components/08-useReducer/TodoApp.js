import React, { useEffect, useReducer } from "react";
import { useForm } from "../../hooks/useForm";
import "../styles/main.css";
import { todoReducer } from "./todoReducer";

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  const [{ description }, handleInputChange, reset] = useForm({
    description: "",
  });

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim().length <= 1) {
      return;
    }

    // nuevo cambio en el estado
    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };

    // action
    const añadirTodo = {
      type: "add",
      payload: newTodo,
    };

    dispatch(añadirTodo);
    reset();
  };

  const deleteTodo = (todoId) => {
    const deleteTs = {
      type: "delete",
      payload: todoId,
    };
    dispatch(deleteTs);
  };

  const toggleTodo = (todoId) => {
    const toggleT = {
      type: "toggle",
      payload: todoId,
    };

    dispatch(toggleT);
  };

  return (
    <>
      <h1>Todo App ({todos.length})</h1>
      <hr />

      <div className="todosMain">
        <div className="todosFlex">
          <ul className="list-group list-group-flush fullWidth">
            {todos.map((todo, i) => (
              <li key={todo.id} className="list-group-item simpleFlex">
                
                <p className={`${todo.done ? 'todo completed': 'todo'}`} onClick={() => toggleTodo(todo.id)}>
                  {i + 1}. {todo.desc}
                </p>
                <button
                  className="btn btn-danger"
                  onClick={() => deleteTodo(todo.id)}
                >
                  Borrar
                </button>
              </li>
            ))}
          </ul>
        </div>
        <div className="agregarColumn">
          <h4>Agregar</h4>
          <br />
          <form className="agregarForm" onSubmit={handleSubmit}>
            <input
              type="text"
              name="description"
              className="form-control"
              placeholder="Nuevo Todo"
              autoComplete="off"
              value={description}
              onChange={handleInputChange}
            />
            <button type="submit" className="btn btn-outline-primary mt-1">
              Agregar
            </button>
          </form>
        </div>
      </div>
    </>
  );
};
