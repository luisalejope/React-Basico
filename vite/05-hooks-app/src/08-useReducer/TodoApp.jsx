
import { TodoList } from "./components/TodoList"
import { TodoAdd } from "./components/TodoAdd"
import { useTodos } from "../hooks/useTodos"

export const TodoApp = () => {

    const {todos, totalTodos, pendingTodos, handleDeleteTodo, handleNewTodo, handleUpdateTodo} = useTodos()

    return (
        <>
            <h1>Todo App: {totalTodos}, <small>Pendientes: {pendingTodos}</small></h1>
            <hr />

            <div className="row">
                <div className="col-7">
                    <TodoList todos={todos} handleDeleteTodo={handleDeleteTodo} handleUpdateTodo={handleUpdateTodo} />
                </div>
                <div className="col-5">
                    <h4>Agregar todo</h4>
                    <hr />
                    <TodoAdd handleSubmit={handleNewTodo} />
                </div>
            </div>

        </>
    )
}
