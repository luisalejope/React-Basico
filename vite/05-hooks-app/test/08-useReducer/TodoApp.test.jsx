import { render } from "@testing-library/react"
import { TodoApp } from "../../src/08-useReducer/TodoApp"
import { useTodos } from "../../src/hooks/useTodos"

jest.mock('../../src/hooks/useTodos')

beforeEach(()=> jest.clearAllMocks())

const todosMock = [
    {
        id: 1,
        desc: 'Correr',
        done: false
    }
]

describe('Pruebas en <TodoApp />', () => {
    useTodos.mockReturnValue({
        todos: todosMock,
        totalTodos: 1,
        pendingTodos: 1, 
        handleDeleteTodo: jest.fn(), 
        handleNewTodo: jest.fn(), 
        handleUpdateTodo: jest.fn()
    });

    test('Debe de corresponder con el snapshot', () => {
        const {container} = render(<TodoApp />)
        expect(container).toMatchSnapshot();
    })


})