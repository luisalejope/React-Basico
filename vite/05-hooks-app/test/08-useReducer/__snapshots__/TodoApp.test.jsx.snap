// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp /> Debe de corresponder con el snapshot 1`] = `
<div>
  <h1>
    Todo App: #
    1
    , 
    <small>
      Pendientes: #
      1
    </small>
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex justify-content-between align-items-center"
        >
          <span
            class=""
          >
            Correr
          </span>
          <button
            class="btn btn-danger"
          >
            Eliminar
          </button>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Agregar todo
      </h4>
      <hr />
      <form>
        <input
          class="form-control"
          name="todoInput"
          type="text"
          value=""
        />
        <button
          class="btn btn-primary mt-2"
          type="submit"
        >
          Agregar
        </button>
      </form>
    </div>
  </div>
</div>
`;
